feat: performance optimization for `changelog` build type

CountRedClaw opened the pull request:
build type: `changelog`
- calculate diffs for whole specifications, split per operations afterwards
- optimize diffs deduplication

build type: `build`
- remove hash calculation for each operation (not used anymore on frontend)

CountRedClaw committed:
test: add operation.deprecated check

CountRedClaw committed:
test: add operation.metadata.originalPath checks

CountRedClaw committed:
style: cleanup

CountRedClaw committed:
test: add operation.deprecatedInfo & operation.deprecatedInPreviousVersions checks

CountRedClaw committed:
refactor: compare documents instead of operations

CountRedClaw committed:
feat: implement documents matching based on operations

CountRedClaw committed:
refactor: simplify documents matching

CountRedClaw committed:
feat: support prefix groups comparison

CountRedClaw committed:
test: add missing cases

CountRedClaw committed:
chore: use api-diff from branch

CountRedClaw committed:
refactor: cleanup

CountRedClaw committed:
refactor: cleanup

CountRedClaw committed:
refactor: simplify comparison logic

CountRedClaw committed:
fix: aggregate missing root-level diffs

CountRedClaw committed:
refactor: cleanup

JayLim2 commented on src/types/external/types.ts:
```diff
@@ -22,3 +29,6 @@ export type FileId = string
 export type VersionId = string | `${string}@${number}`
 export type OperationId = string
+
+export type WithDiffMetaRecord<T> = T & {[DIFF_META_KEY]?: DiffMetaRecord}
```
T extends Record<PropertyKey, unknown> to prevent mistakes

CountRedClaw replied:
not suitable for OpenAPIV3.PathsObject, OpenAPIV3.Document and OpenAPIV3.ServerObject

JayLim2 commented on test/projects/changelog/change-method/after.yaml:
```diff
@@ -0,0 +1,13 @@
+  version: 0.1.0
+paths:
+  /api/v1/path1:
+    get:
```
nothing changed

JayLim2 commented on test/projects/changelog/change-method/before.yaml:
```diff
@@ -0,0 +1,13 @@
+  version: 0.1.0
+paths:
+  /api/v1/path1:
+    get:
```
nothing changed (before)

CountRedClaw replied:
content type changed from application/xml to application/json

JayLim2 commented on test/projects/changelog/documents-matching/before/spec2.yaml:
```diff
@@ -0,0 +1,34 @@
+openapi: 3.0.0
+info:
```
isn't it redundant to have `spec2`? it is almost the same that `spec` is, just swapped, isn't it?

CountRedClaw committed:
test: add missing cases

CountRedClaw commented on src/apitypes/rest/rest.changes.ts:
```diff
@@ -14,80 +14,199 @@
-  const olnyBreaking = diffResult.diffs.filter((diff) => diff.type === breaking)
-  if (olnyBreaking.length > 0 && previous?.operationId) {
-    await reclassifyBreakingChanges(previous.operationId, diffResult.merged, olnyBreaking, ctx)
+  ) as { merged: OpenAPIV3.Document; diffs: Diff[] }
```
we pass only published openapi docs to the apiDiff here, but it returns merged: unknown, so I think that this cast is justified

CountRedClaw commented on src/apitypes/rest/rest.changes.ts:
```diff
@@ -14,80 +14,199 @@
+      await reclassifyBreakingChanges(previous?.operationId, merged, operationDiffs, ctx)
+
+      changedOperations.push(createOperationChange(apiType, operationDiffs, previous, current, currGroupSlug, prevGroupSlug, currentGroup, previousGroup))
+      getOperationTags(current ?? previous).forEach(tag => tags.add(tag))
```
yes, but in this case we collected the diffs from the current operation only - https://github.com/Netcracker/qubership-apihub-api-processor/blob/19be6cab7b9ab04634bb83fdb2c8acb6cdf0549f/src/components/compare/compare.operations.ts#L331

CountRedClaw committed:
fix: review remarks

CountRedClaw committed:
refactor

CountRedClaw committed:
perf: add data for measurements

CountRedClaw committed:
perf: add data for measurements

CountRedClaw committed:
Merge remote-tracking branch 'origin/develop' into feature/performance-optimization

# Conflicts:
#	package.json

CountRedClaw committed:
chore: use api-diff from branch

CountRedClaw committed:
fix: remove prefix from servers case

CountRedClaw committed:
feat: collect all deep changes in components.securitySchemes

CountRedClaw committed:
fix: servers and security diffs collecting

CountRedClaw committed:
fix: review remarks

CountRedClaw committed:
fix: review remarks

CountRedClaw committed:
test: path change due to moving a prefix from server.url to path classified as annotation

CountRedClaw committed:
fix: duplicated tags

CountRedClaw committed:
refactor

CountRedClaw committed:
fix: linter

CountRedClaw committed:
cleanup

CountRedClaw committed:
test: fix data

CountRedClaw committed:
fix: ci linter

CountRedClaw committed:
fix: make fileResolver optional

CountRedClaw committed:
fix: graphql operations handling

CountRedClaw committed:
fix: prefix groups changelog calculation when prefix is specified in root servers[0].url

CountRedClaw committed:
test: add unsupported cases

b41ex committed:
fix: use existing document pair (if available) to calculate diffs for added/removed operations

results in up to 3 times performance improvements for typical cases

CountRedClaw committed:
perf: filter out referentially equal diffs before diffId calculation

b41ex committed:
refactor: fix expected result for test according to new behaviour in api-diff

b41ex committed:
fix: use recursive aggregateDiff function function (many-folds perfromance improvement)

b41ex committed:
fix: optimize diff deduplication- use reference identity when diffs are from the same document pair

b41ex committed:
fix: optimize operationsChanges deduplication for typical case

b41ex committed:
chore: set feature branch version for depdendencies

b41ex committed:
chore: merge remote-tracking branch 'refs/remotes/origin/feature/performance-optimization' into feature/performance-optimization

b41ex committed:
chore: merge branch 'develop' into feature/performance-optimization

b41ex committed:
chore: merge develop

b41ex committed:
chore: fix linter errors

b41ex committed:
chore: fix linter errors

b41ex committed:
chore: fix linter errors

b41ex committed:
chore: disable GraphQL linting

b41ex committed:
chore: remove unused imports

b41ex committed:
fix: extra leading dash in the the operation id when comparing prefix groups

makeev-pavel commented on src/apitypes/rest/rest.changes.ts:
```diff
@@ -14,84 +14,198 @@
+      // todo if there were actually servers here, we wouldn't have handle it, add a test
+      const previousBasePath = getOperationBasePath(methodData?.servers || pathData?.servers || prevDocData.servers || [])
+      const currentBasePath = getOperationBasePath(methodData?.servers || pathData?.servers || currDocData.servers || [])
+      const prevNormalizedOperationId = slugify(`${normalizePath(previousBasePath + path)}-${inferredMethod}`, [], IGNORE_PATH_PARAM_UNIFIED_PLACEHOLDER)
```
142 - 143 
maybe it should be made into a separate function?

makeev-pavel commented on src/apitypes/rest/rest.types.ts:
```diff
@@ -70,3 +70,16 @@ export interface RestOperationContext {
   [REST_KIND_KEY]?: string
 }
+
+// export type Merged = {
```
if you need it, then add a TODO or a description of why

b41ex committed:
fix: operation matching when operation base path moved between path and path item servers object when comparing prefix groups

b41ex committed:
refactor: rename test to avoid problems with selective run

b41ex committed:
fix: incorrect calculation of operation ids in prefix groups

b41ex committed:
fix: enforce group prefix validation

b41ex committed:
refactor: skip useless test

b41ex committed:
fix: add pahth parameter rename test for prefix groups

b41ex committed:
fix return full operation ids when comparing prefix groups

b41ex committed:
refactor: reuse group prefix validation function

b41ex committed:
refactor: export createCopyWithPrefixGroupOperationsOnly for reuse in UI

b41ex committed:
refactor: modify createCopyWithPrefixGroupOperationsOnly to keep copy as close to source as possible

b41ex committed:
refactor: export RestOperationData to use in UI

b41ex committed:
refactor: fix code formatting

b41ex committed:
refactor: use existing isValidHttp method

b41ex committed:
refactor: extract method to calculateNormalizedOperationId

b41ex committed:
refactor: remove obsolete comment

b41ex committed:
docs: add comments

b41ex committed:
refactor: add unit test for removeRedundantPartialPairs

b41ex committed:
docs: rename test group to emphasize it is for the changelog build type

b41ex committed:
refactor: use extractOperationBasePath from api-diff to avoid code duplication

b41ex committed:
refactor: switch between and after samples for add-prefix-to-server to avoid confusion

b41ex committed:
refactor: add unit tests for dedupeTuples function

b41ex committed:
fix: collect openapi version changes for operations

b41ex committed:
fix: add missing samples

b41ex committed:
fix: add validation of api-processor version for previous and current package versions when building changelog

b41ex committed:
fix: do not call resolvers for falsy versions when building a changelog

b41ex committed:
fix: do not allow to publish specifications with operations with empty path parameter name

b41ex committed:
fix: always expect apiProcessorVersion for resolved versions

b41ex committed:
fix: should fail publication of package version with duplicated operations

b41ex committed:
refactor: add test for path parameter name path fragment collision

b41ex committed:
refactor: add a test checking that resolvers are not called with empty version when building dashboard with added/removed packages

b41ex committed:
fix: change spies to actually work

b41ex committed:
refactore: add outdated api-processor test for dashboard changelog built

b41ex committed:
refactor: add test for outdated api-processor validation when building package changelog

vOrigins committed:
fix: improve error handling for duplicate operationIds

- Enhance error messages to specify if duplicates are within a single document or across multiple documents.
- Add tests for both scenarios: duplicates within a document and across documents.
- Update document processing logic to detect and throw detailed errors for duplicate operationIds.

vOrigins committed:
fix: improve duplicate operationId detection

- Refactor logic to handle duplicate operationIds after processing all operations within a document.
- Update error messages to provide detailed information about duplicates.
- Add tests to validate detection of multiple duplicates and three or more operations with the same operationId.

b41ex committed:
refactor: use before/after normalized values from symbols (#46)

b41ex committed:
fix: collect only security and security schema changes relevant for the operation (#45)

fix: collect only security and security schema changes relevant for the operation
fix: only include root security and security schemes relevant for operation when creating single operation specification

b41ex committed:
chore: merge branch 'develop' into feature/performance-optimization

b41ex committed:
chore: merge branch 'develop' into feature/performance-optimization

b41ex commented:
Will be merged inhttps://github.com/Netcracker/qubership-apihub-api-processor/pull/47
